pkg/ottl/ottlfuncs/func_scale.go

```diff
@@ -0,0 +1,132 @@
+package ottlfuncs
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+	"go.opentelemetry.io/collector/pdata/pmetric"
+)
+
+type ScaleArguments[K any] struct {
+	Value      ottl.Getter[K]
+	Multiplier float64
+}
+
+func NewScaleFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("Scale", &ScaleArguments[K]{}, createScaleFunction[K])
+}
+
+func createScaleFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ScaleArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ScaleFactory args must be of type *ScaleArguments[K]")
+	}
+
+	return Scale(args.Value, args.Multiplier)
+}
+
+func Scale[K any](value ottl.Getter[K], multiplier float64) (ottl.ExprFunc[K], error) {
+	return func(ctx context.Context, tCtx K) (any, error) {
+		get, err := value.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		switch get.(type) {
+		case float64:
+			return get.(float64) * multiplier, nil
+		case int64:
+			return float64(get.(int64)) * multiplier, nil
+		case pmetric.NumberDataPointSlice:
+			scaledMetric := pmetric.NewNumberDataPointSlice()
+			get.(pmetric.NumberDataPointSlice).CopyTo(scaledMetric)
+			scaleMetric(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.HistogramDataPointSlice:
+			scaledMetric := pmetric.NewHistogramDataPointSlice()
+			get.(pmetric.HistogramDataPointSlice).CopyTo(scaledMetric)
+			scaleHistogram(scaledMetric, multiplier)
+			return scaledMetric, nil
+		case pmetric.ExponentialHistogramDataPointSlice:
+			scaledMetric := pmetric.NewExponentialHistogramDataPointSlice()
+			get.(pmetric.ExponentialHistogramDataPointSlice).CopyTo(scaledMetric)
+			scaleExponentialHistogram(scaledMetric, multiplier)
+			return scaledMetric, nil
+		default:
+			return nil, errors.New("unsupported data type")
+		}
```

Comment:
```suggestion
		untypedVal, err := value.Get(ctx, tCtx)
		if err != nil {
			return nil, err
		}

		switch val := untypedVal.(type) {
		case float64:
			return val * multiplier, nil
		case int64:
			return float64(val) * multiplier, nil
		case pmetric.NumberDataPointSlice:
			scaledMetric := pmetric.NewNumberDataPointSlice()
			val.CopyTo(scaledMetric)
			scaleMetric(scaledMetric, multiplier)
			return scaledMetric, nil
		case pmetric.HistogramDataPointSlice:
			scaledMetric := pmetric.NewHistogramDataPointSlice()
			val.CopyTo(scaledMetric)
			scaleHistogram(scaledMetric, multiplier)
			return scaledMetric, nil
		case pmetric.ExponentialHistogramDataPointSlice:
			scaledMetric := pmetric.NewExponentialHistogramDataPointSlice()
			val.CopyTo(scaledMetric)
			scaleExponentialHistogram(scaledMetric, multiplier)
			return scaledMetric, nil
		default:
			return nil, errors.New("unsupported data type")
		}
```
Go provides a variable with the proper type inside each `case` statement if you create one, which I think cleans this up a bit. We should also probably go with a local variable name other than `get` since that's a verb.

Reply:
thanks for the suggestion, that tidies it up a bit